sdks/java/extensions/ml/src/main/java/org/apache/beam/sdk/extensions/ml/DLPDeidentifyText.java

```diff
@@ -0,0 +1,267 @@
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.dlp.v2.DlpServiceClient;
+import com.google.privacy.dlp.v2.ContentItem;
+import com.google.privacy.dlp.v2.DeidentifyConfig;
+import com.google.privacy.dlp.v2.DeidentifyContentRequest;
+import com.google.privacy.dlp.v2.DeidentifyContentResponse;
+import com.google.privacy.dlp.v2.FieldId;
+import com.google.privacy.dlp.v2.InspectConfig;
+import com.google.privacy.dlp.v2.ProjectName;
+import com.google.privacy.dlp.v2.Table;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionView;
+
+/**
+ * A {@link PTransform} connecting to Cloud DLP (https://cloud.google.com/dlp/docs/libraries) and
+ * deidentifying text according to provided settings. The transform supports both CSV formatted
+ * input data and unstructured input.
+ *
+ * <p>If the csvHeader property is set and a sideinput with CSV headers is added to the PTransform,
+ * csvDelimiter also should be set, else the results will be incorrect. If csvHeader is neither set
+ * nor passed as sideinput, input is assumed to be unstructured.
+ *
+ * <p>Either deidentifyTemplateName (String) or deidentifyConfig {@link DeidentifyConfig} need to be
+ * set. inspectTemplateName and inspectConfig ({@link InspectConfig} are optional.
+ *
+ * <p>Batch size defines how big are batches sent to DLP at once in bytes.
+ *
+ * <p>The transform consumes {@link KV} of {@link String}s (assumed to be filename as key and
+ * contents as value) and outputs {@link KV} of {@link String} (eg. filename) and {@link
+ * DeidentifyContentResponse}, which will contain {@link Table} of results for the user to consume.
+ */
+@Experimental
+@AutoValue
+public abstract class DLPDeidentifyText
+    extends PTransform<
+        PCollection<KV<String, String>>, PCollection<KV<String, DeidentifyContentResponse>>> {
+
+  public static final Integer DLP_PAYLOAD_LIMIT_BYTES = 524000;
+
+  /** @return Template name for data inspection. */
+  @Nullable
+  public abstract String inspectTemplateName();
+
+  /** @return Template name for data deidentification. */
+  @Nullable
+  public abstract String deidentifyTemplateName();
+
+  /**
+   * @return Configuration object for data inspection. If present, supersedes the template settings.
+   */
+  @Nullable
+  public abstract InspectConfig inspectConfig();
+
+  /** @return Configuration object for deidentification. If present, supersedes the template. */
+  @Nullable
+  public abstract DeidentifyConfig deidentifyConfig();
+
+  /** @return List of column names if the input KV value is a CSV formatted row. */
```

Comment:
I'm getting a bit confused at when a header is present and when one isn't now. When building the table for the request to the cloud API, if no header is present, we set one to "value" by default. If that column header ends up in the response, should this method also return "value" when set to null?

If that is the case, this needs to be clear in the comments somewhere.

Reply:
I don't think this should return "value" if there's no header from the user, this property is for user-defined headers only and I don't think there's need to return the defaults anyway. We'd need to evaluate existence of the sideinput in the internal DoFn anyway, so there wouldn't be much gain in terms of code complexity anyway.

Reply:
OK, sounds good.